spot.go

```diff
@@ -2,65 +2,90 @@ package main
 
 import (
 	"context"
-	"io/ioutil"
-	"net/http"
+	"errors"
+	"strings"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+
 	log "github.com/sirupsen/logrus"
 )
 
-const (
-	metadataURLTerminationTime = "http://169.254.169.254/latest/meta-data/spot/termination-time"
-	terminationTransition      = "ec2:SPOT_INSTANCE_TERMINATION"
-	terminationTimeFormat      = "2006-01-02T15:04:05Z"
-)
+// NewSpotListener ...
+func NewSpotListener(instanceID string, metadata *ec2metadata.EC2Metadata) *SpotListener {
+	return &SpotListener{
+		listenerType: "spot",
+		instanceID:   instanceID,
+		metadata:     metadata,
+	}
+}
 
-func pollSpotTermination(ctx context.Context) chan time.Time {
-	ch := make(chan time.Time)
+// SpotListener ...
+type SpotListener struct {
+	listenerType string
+	instanceID   string
+	metadata     *ec2metadata.EC2Metadata
+}
 
-	log.Debugf("Polling metadata service for spot termination notices")
+// Type returns a string describing the listener type.
+func (l *SpotListener) Type() string {
+	return l.listenerType
+}
 
-	go func() {
-		// Close channel before returning since this (goroutine) is the sending side.
-		defer close(ch)
-		retry := time.NewTicker(time.Second * 5).C
-	Loop:
-		for {
-			select {
-			case <-ctx.Done():
-				break Loop
-			case <-retry:
-				res, err := http.Get(metadataURLTerminationTime)
-				if err != nil {
-					log.WithError(err).Info("Failed to query metadata service")
-					continue
-				}
+// Start the spot termination notice listener.
+func (l *SpotListener) Start(ctx context.Context, notices chan<- TerminationNotice) error {
+	if !l.metadata.Available() {
+		return errors.New("ec2 metadata is not available")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.NewTicker(time.Second * 5).C:
+			log.Debugf("Polling ec2 metadata for spot termination notices")
 
-				// We read the body immediately so that we can close the body in one place
-				// and still use 'continue' if any of our conditions are false.
-				body, err := ioutil.ReadAll(res.Body)
-				res.Body.Close()
-				if err != nil {
-					log.WithError(err).Info("Failed to read response from metadata service")
+			out, err := l.metadata.GetMetadata("spot/termination-time")
+			if err != nil {
+				if e, ok := err.(awserr.Error); ok && strings.Contains(e.OrigErr().Error(), "404") {
```

Comment:
Btw @lox - I could not find a better solution for this, the error types returned from the `ec2metadata` service are less informative than the other services. This solves all the false positive errors you mentioned, but needs to be tested with an actual spot termination to make sure that it works.

Reply:
Yeah, that's what I ended up with. I just used a longer pattern match on the error.Error(). Should be fine.